File: iteration2/src/CommandLineInterface/CLIAdvisor.java
package iteration2.src.CommandLineInterface;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import iteration2.src.controllers.AdvisorController;
import iteration2.src.enums.Color;
import iteration2.src.enums.CourseStatus;
import iteration2.src.models.Advisor;
import iteration2.src.models.SelectedCourse;
import iteration2.src.models.Student;
import iteration2.src.utils.Util;

public class CLIAdvisor {

    private Advisor advisor;
    private AdvisorController advisorController;
    Scanner input;

    public CLIAdvisor(Advisor advisor) {
        this.advisor = advisor;
        advisorController = new AdvisorController(advisor);
    }

    public void menuPage() {

        boolean isInvalid = false;
        input = new Scanner(System.in);
        AdvisorController advisorController = new AdvisorController(advisor);
        while (true) {
            Util.clearScreen();
            if (advisorController.getNotification() != null && advisorController.getNotification().size() > 0) {
                Util.paintTextln(" Notifications\n " +
                        "**************", Color.YELLOW);
                for (String string : advisorController.getNotification())
                    Util.paintTextln(string, Color.RED);
                advisorController.clearNotifications();
            }

            System.out.println("\n\n");

            if (!isInvalid) {
                isInvalid = false;
                // while(true) {
                System.out.println(" Menu ");
                System.out.println("********");
                System.out.println("    1. Student List");
                System.out.println("    2. Log out");
            } else {
                isInvalid = false;
            }
            boolean shouldQuit = false;
            String choice = input.nextLine();

            if (choice.equals("1")) {
                List<Student> students = advisorController.getStudentListOrderByStatus();
                shouldQuit = showStudentPage(students);

                if (shouldQuit) {
                    break;
                } else {
                    continue;
                }
            } else if (choice.equals("2")) {
                advisorController.logOut();
                break;
            } else {
                Util.sendFeedback("Invalid choice", Color.RED);
                isInvalid = true;
            }
        }
    }

    public boolean showStudentPage(List<Student> students) {
        boolean isInvalid = false;
        while (true) {
            if (!isInvalid) {
                isInvalid = false;

                System.out.println(" Student List");
                System.out.println("****************");
                /*
                 * System.out.println("   Number      Name");
                 * System.out.println("   ------      -----");
                 */
                System.out.printf("%10s    %30s\n", "Number", "Name");
                for (int i = 0; i < students.size(); i++) {

                    System.out.printf("%d.  %10s    %30s\n", i + 1, students.get(i).getUserId(),
                            students.get(i).getFirstName() + " " + students.get(i).getLastName());
                    /*
                     * System.out.println((i + 1) + ". " + students.get(i).getUserId() + "    "
                     * + students.get(i).getFirstName() + " " + students.get(i).getLastName());
                     */
                }

                System.out.println("Select Student:");
                System.out.println("Press b to back");
                System.out.println("Press q to quit");
            } else {
                isInvalid = false;
            }
            List<SelectedCourse> selectedCoursesThatStatusIsPending = new ArrayList<SelectedCourse>();

            String choice = input.nextLine();
            try {
                int choice2 = Integer.parseInt(choice);
                if (choice2 > 0 && choice2 <= students.size()) {
                    for (int i = 0; i < students.get(choice2 - 1).getSelectedCourses().size(); i++) {
                        if (students.get(choice2 - 1).getSelectedCourses().get(i).getStatus() == CourseStatus.PENDING) {
                            selectedCoursesThatStatusIsPending
                                    .add(students.get(choice2 - 1).getSelectedCourses().get(i));
                        }
                    }
                    boolean shouldQuit = coursesOfStudentPage(students.get(choice2 - 1));
                    if (shouldQuit) {
                        return true;
                    }
                } else {
                    Util.sendFeedback("Invalid choice. Try again", Color.RED);
                    isInvalid = true;
                }
            } catch (NumberFormatException e) {
                if (choice.equals("b")) {
                    return false;
                } else if (choice.equals("q")) {
                    advisorController.logOut();
                    return true;
                } else {
                    Util.sendFeedback("Invalid choice. Try again", Color.RED);
                    isInvalid = true;
                }
            }
        }
    }

    public boolean coursesOfStudentPage(Student student) {

        boolean isInvalid = false;
        List<SelectedCourse> courses = student.fetchSelectedCoursesForAdvisor();
        while (true) {
            if (!isInvalid) {
                isInvalid = false;

                Util.clearScreen();

                System.out.println(
                        " Student: " + student.getFirstName() + " " + student.getLastName() + "'s TimeTable \n");
                Util.printTimeTable(student.createTimeTable());

                System.out.println(" Courses Of The Student ");
                System.out.println("***************************");
                /*
                 * System.out.println("   Code         Number        Section       Status ");
                 * System.out.println("  ------        ------         ------       ------ ");
                 */
                System.out.printf("    %10s    %50s    %15s    %15s\n", "Code", "Name", "Section", "Status");
                System.out.printf("    %10s    %50s   %15s     %15s\n", "-----", "------", "--------", "------");

                for (int i = 0; i < courses.size(); i++) {
                    /*
                     * System.out.println((i + 1) + ". " +
                     * courses.get(i).getCourse().getCourseCode() + "    "
                     * + courses.get(i).getCourse().getCourseName() + "    "
                     * + courses.get(i).getCourseSection().getSectionCode() + "    " +
                     * courses.get(i).getStatus());
                     */
                    System.out.printf("%d.  %10s    %50s    %15s             ", i + 1,
                            courses.get(i).getCourse().getCourseCode(),
                            courses.get(i).getCourse().getCourseName(),
                            courses.get(i).getCourseSection().getSectionCode());

                    if (courses.get(i).getStatus() == CourseStatus.APPROVED)
                        Util.paintText(courses.get(i).getStatus() + "\n", Color.GREEN);
                    else if (courses.get(i).getStatus() == CourseStatus.PENDING)
                        Util.paintText(courses.get(i).getStatus() + "\n", Color.YELLOW);
                    else if (courses.get(i).getStatus() == CourseStatus.DENIED)
                        Util.paintText(courses.get(i).getStatus() + "\n", Color.RED);
                    else
                        System.out.print(courses.get(i).getStatus() + "\n");
                }

                System.out.println("Press b to back");
                System.out.println("Press q to quit");
                System.out.println("Select Course: ");
            } else {
                isInvalid = false;
            }
            String choice = input.nextLine();
            try {
                int choiceInt = Integer.parseInt(choice);
                if (choiceInt > 0 && choiceInt <= courses.size()) {
                    System.out.println("Press a to approve");
                    System.out.println("Press d to deny");

                    String choice2 = input.nextLine();

                    if (choice2.equals("a")) {
                        advisorController.approveCourse(student, courses.get(choiceInt - 1));
                        System.out.println();
                    } else if (choice2.equals("d")) {
                        advisorController.denyCourse(student, courses.get(choiceInt - 1));
                    } else if (choice.equals("b")) {
                        return false;
                    } else if (choice.equals("q")) {
                        advisorController.logOut();
                        return true;
                    } else {
                        Util.sendFeedback("Invalid choice. Try again", Color.RED);
                        isInvalid = true;
                    }
                } else {
                    ;
                    Util.sendFeedback("Invalid choice!", Color.RED);
                }
            } catch (NumberFormatException e) {
                if (choice.equals("b")) {
                    return false;
                } else if (choice.equals("q")) {
                    advisorController.logOut();
                    return true;
                } else {
                    Util.sendFeedback("Invalid choice. Try again", Color.RED);
                    isInvalid = true;
                }
            }

        }

    }
}

File: iteration3/src/CommandLineInterface/CLIAdvisor.py
from controllers.AdvisorController import AdvisorController
from enums.CourseStatus import CourseStatus
from interfaces.Color import Color
from models.SelectedCourse import SelectedCourse
from models.Student import Student
from utils.Util import Util
from models.Constraint import Constraint

class CLIAdvisor(object):

     def __init__(self, advisorController):
        self._advisorController = advisorController


     def menuPage(self):
          while True:

               Util.clearScreen()

               if self._advisorController.getNotifications() is not None:
                    Util.paintTextln(" Notifications\n " +
                                "**************", Color.YELLOW)
                    for string in self._advisorController.getNotifications():
                         Util.paintTextln(string, Color.RED)
                    self._advisorController.clearNotifications()

               print("\n\n")
               print(" Menu ")
               print("********")
               print("    1. Student List")
               print("    2. Log out")
               
               
               shouldQuit = False

               choice: str = input()

               if choice == "1":
                    students = self._advisorController.getStudentListOrderByStatus()
                    Util.clearScreen()
                    shouldQuit = self._showStudentPage(students)

                    if(shouldQuit):
                         break
                    else:
                         continue
               elif choice == "2":
                    self._advisorController.logOut()
                    break
               else:
                    Util.sendFeedback("Invalid choice", Color.RED)
                    isInvalid = True

     def _showStudentPage(self,students):
          isInvalid = False
          while True:
               
               print("Student List")
               print("************")
               print(f"{('Number'):>10}    {('Name'):>30}")
                    
               for i in range(len(students)):
                    print(f"{i + 1}  {students[i].getUserId():>10}    {students[i].getFirstName() + ' '  + students[i].getLastName():>30}")
               
               
               print("\nPress b to back")
               print("Press q to quit\n")   
               print("Select Student:")       
               
               selectedCoursesThatStatusIsPending = []

               choice = input()

               try: 
                    if(choice == "b"):
                         return False
                    elif(choice == "q"):
                         self._advisorController.logOut()
                         return True
                    elif(Util.validateNumber(choice, students)):
                         for i in range(len(students[int(choice)-1].getSelectedCourses())):
                                   if((students[int(choice)-1].getSelectedCourses())[i].getStatus() == CourseStatus.PENDING):
                                        selectedCoursesThatStatusIsPending.append((students[int(choice)-1].getSelectedCourses())[i])
                              
                         shouldQuit = self._coursesOfStudentPage(students[int(choice)-1])         
                         if(shouldQuit):
                              return True
                    
                    else:
                         Util.sendFeedback("Invalid choice. Try again.",Color.RED)
                         isInvalid = True
               except Exception as e:
                    Util.sendFeedback(str(e), Color.RED)
   

     def _coursesOfStudentPage(self,student):

          isInvalid = False    
          courses = student.fetchSelectedCoursesForAdvisor()
          isFinalizable = False
          
          while True:
               
               Util.clearScreen()

               print(" Student: " + student.getFirstName() + " " + student.getLastName() + "'s TimeTable \n")
               Util.printTimeTable(student.createTimeTable())

               print("\n\n Courses Of The Student ")
               print("***************************")

               print(f"    {'Code':>10}    {'Name':>50}    {'Section':>15}    {'Status':>15}")
               print(f"    {'-----':>10}    {'------':>50}    {'--------':>15}    {'------':>15}")

               for i in range(len(courses)):
                    print(f"{i+1}  {courses[i].getCourse().getCourseCode():>10}    {courses[i].getCourse().getCourseName():>50}    {courses[i].getCourseSection().getSectionCode():>15}   " , end="           ")

                    if courses[i].getStatus() == CourseStatus.APPROVED:
                         Util.paintText(courses[i].getStatus() + "\n", Color.GREEN)
                    elif courses[i].getStatus() == CourseStatus.PENDING:
                         Util.paintText(courses[i].getStatus() + "\n", Color.YELLOW)
                    elif courses[i].getStatus() == CourseStatus.DENIED:
                         Util.paintText(courses[i].getStatus() + "\n", Color.RED)
                    elif courses[i].getStatus() == CourseStatus.APPROVED_FINALIZED:
                         Util.paintText("APPROVED FINALIZED" + "\n", Color.GREEN)
                    elif courses[i].getStatus() == CourseStatus.DENIED_FINALIZED:
                         Util.paintText("DENIED FINALIZED" + "\n", Color.RED)
                    else:
                         print(courses[i].getStatus() + "\n")

               print("\nPress s to show timetable")
               print("Press f to finalize registration")
               print("Press b to back")
               print("Press q to quit\n")
               print("Select Course: ")
              
              

              
               choice = input()

               try:
                    if choice == "b":
                         return False
                    elif choice=="q":
                         self._advisorController.logOut()
                         return True
                    elif choice == "f":
                         if(len(student.getSelectedCourses()) == 0):
                              print("Student has not selected any course yet. Are you sure to finalize registration? (y/n)")
                              choice2 = input()
                              if choice2 == "y":
                                   self._advisorController.finalizeRegistration(student)
                                   Util.sendFeedback("Registration is finalized.", Color.GREEN)
                              elif choice2 == "n":
                                   continue
                              else:
                                   raise Exception("Invalid choice")
                         else:
                              self._advisorController.finalizeRegistration(student)
                              Util.sendFeedback("Registration is finalized.", Color.GREEN)
                    elif choice == "s":
                         Util.printTimeTable(student.createTimeTable())
                    elif(Util.validateNumber(choice, courses)):
                              print("Press a to approve")
                              print("Press d to deny")
                         
                              choice2 = input()

                              if choice2 == "a" :                                        
                                        self._advisorController.approveCourse(student, courses[int(choice) - 1])
                                        Util.sendFeedback("Course is approved", Color.GREEN)
                              elif choice2 == "d":                                   
                                             self._advisorController.denyCourse(student, courses[int(choice) - 1])
                                             Util.sendFeedback("Course is denied", Color.GREEN)
                              else:
                                   raise Exception("Invalid choice. Try again")    
                    else:
                         raise Exception("Invalid choice. Try again")
               except Exception as e:
                    Util.sendFeedback(str(e), Color.RED)
                    